File: crates/compiler/frontend/native/typescript/js/conversational.ts
// Conversational Agent Program composition on public five-op authoring APIs.

import {
  AgentProgram,
  type ContextEdge,
  type HookBinding,
  type ImportedProgram,
} from "./agent-program.ts";

export type TurnSpec = {
  model_node_id?: string;
  capability_node_id?: string;
  await_node_id?: string;
  model_target_ref?: string;
  capability_ref?: string;
  event_ref?: string;
};

export type SpecialistComposition = {
  import_ref: ImportedProgram;
  new_node_id?: string;
  invoke_node_id?: string;
};

export class ConversationalAgent extends AgentProgram {
  static readonly LOOP_REGION_SUFFIX = "loop.turn";
  static readonly RETURN_REGION_ID = "region.return";

  readonly loopRegionId: string;
  private specialist: SpecialistComposition | undefined;
  private readonly sourceFile: string;

  constructor(opts: {
    program_id: string;
    input_type_ref: string;
    output_type_ref: string;
    context_type_ref: string;
    entrypoint?: string;
    source_language?: string;
  }) {
    super({
      program_id: opts.program_id,
      entrypoint: opts.entrypoint,
      input_type_ref: opts.input_type_ref,
      output_type_ref: opts.output_type_ref,
      has_default_context: true,
      context_type_ref: opts.context_type_ref,
      source_language: opts.source_language,
    });
    this.loopRegionId = `region.${ConversationalAgent.LOOP_REGION_SUFFIX}`;
    this.sourceFile = `${opts.program_id}.${opts.source_language === "typescript" ? "ts" : "py"}`;
  }

  withSpecialist(composition: SpecialistComposition): this {
    this.specialist = composition;
    return this;
  }

  defineTurn(
    turn: TurnSpec = {},
    opts: { hooks?: HookBinding[] } = {},
  ): this {
    const modelNodeId = turn.model_node_id ?? "node.turn.model";
    const capabilityNodeId = turn.capability_node_id ?? "node.turn.tool";
    const awaitNodeId = turn.await_node_id ?? "node.turn.await";
    const modelTargetRef = turn.model_target_ref ?? "model.default";
    const eventRef = turn.event_ref ?? "event.turn.input";

    if (this.specialist !== undefined) {
      this.importProgram(this.specialist.import_ref);
    }
    this.builder.modelCall(modelNodeId, modelTargetRef);
    if (turn.capability_ref !== undefined) {
      this.builder.capabilityInvoke(capabilityNodeId, turn.capability_ref);
    }
    if (this.specialist !== undefined) {
      this.builder.programNew(this.specialist.new_node_id ?? "node.specialist.new");
      this.builder.programInvoke(this.specialist.invoke_node_id ?? "node.specialist.invoke");
    }
    this.builder.awaitEvent(awaitNodeId, eventRef);
    this.builder.region(this.loopRegionId, "loop");
    this.builder.returnRegion(ConversationalAgent.RETURN_REGION_ID);

    const contextSink =
      turn.capability_ref !== undefined ? capabilityNodeId : awaitNodeId;
    this.contextFlow({
      from_node: modelNodeId,
      to_node: contextSink,
      context_type_ref: this.contextTypeRef ?? "",
    });

    for (const hook of opts.hooks ?? []) {
      this.bindHook(hook);
    }

    this.builder.modelRequirement(modelTargetRef);
    if (turn.capability_ref !== undefined) {
      this.builder.capabilityRequirement(turn.capability_ref);
    }
    const sourceNodes: Array<[string, string]> = [[modelNodeId, "model.call"]];
    if (turn.capability_ref !== undefined) {
      sourceNodes.push([capabilityNodeId, "capability.invoke"]);
    }
    if (this.specialist !== undefined) {
      sourceNodes.push([
        this.specialist.new_node_id ?? "node.specialist.new",
        "program.new",
      ]);
      sourceNodes.push([
        this.specialist.invoke_node_id ?? "node.specialist.invoke",
        "program.invoke",
      ]);
    }
    sourceNodes.push([awaitNodeId, "await.event"]);
    sourceNodes.forEach(([nodeId, annotation], index) => {
      this.builder.nodeSpan(nodeId, this.sourceFile, index + 1, annotation);
    });

    return this.annotateRegion(this.loopRegionId, "conversational_loop");
  }
}

File: crates/compiler/frontend/python/apxm_program/conversational.py
"""Conversational Agent Program composition on public five-op authoring APIs."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from .agent_program import AgentProgram, ContextEdge, HookBinding, ImportedProgram


@dataclass(frozen=True, slots=True)
class TurnSpec:
    """One authored conversational Turn inside a structural loop region."""

    model_node_id: str = "node.turn.model"
    capability_node_id: str = "node.turn.tool"
    await_node_id: str = "node.turn.await"
    model_target_ref: str = "model.default"
    capability_ref: Optional[str] = None
    event_ref: str = "event.turn.input"


@dataclass(frozen=True, slots=True)
class SpecialistComposition:
    """Typed specialist Program Instance composition inside a Turn."""

    import_ref: ImportedProgram
    new_node_id: str = "node.specialist.new"
    invoke_node_id: str = "node.specialist.invoke"


class ConversationalAgent(AgentProgram):
    """Compose a conversational loop from public Agent Program constructs only.

    The loop is a structural region annotated as a conversational loop. Each Turn
    authors explicit `model.call`, optional `capability.invoke`, optional
    `program.new`/`program.invoke`, and `await.event` park/resume semantics. No
    retired operation, hidden runtime loop, or compiler branch is introduced.
    """

    LOOP_REGION_SUFFIX = "loop.turn"
    RETURN_REGION_ID = "region.return"

    def __init__(
        self,
        *,
        program_id: str,
        input_type_ref: str,
        output_type_ref: str,
        context_type_ref: str,
        entrypoint: str = "run",
        source_language: str = "python",
    ) -> None:
        super().__init__(
            program_id=program_id,
            entrypoint=entrypoint,
            input_type_ref=input_type_ref,
            output_type_ref=output_type_ref,
            has_default_context=True,
            context_type_ref=context_type_ref,
            source_language=source_language,
        )
        self._loop_region_id = f"region.{self.LOOP_REGION_SUFFIX}"
        self._specialist: Optional[SpecialistComposition] = None
        self._source_file = f"{program_id}.{'ts' if source_language == 'typescript' else 'py'}"

    @property
    def loop_region_id(self) -> str:
        return self._loop_region_id

    def with_specialist(self, composition: SpecialistComposition) -> "ConversationalAgent":
        self._specialist = composition
        return self

    def define_turn(
        self,
        turn: TurnSpec | None = None,
        *,
        hooks: tuple[HookBinding, ...] = (),
    ) -> "ConversationalAgent":
        spec = turn or TurnSpec()
        builder = self.builder
        if self._specialist is not None:
            self.import_program(self._specialist.import_ref)
        builder.model_call(spec.model_node_id, model_target_ref=spec.model_target_ref)
        if spec.capability_ref is not None:
            builder.capability_invoke(spec.capability_node_id, capability_ref=spec.capability_ref)
        if self._specialist is not None:
            builder.program_new(self._specialist.new_node_id)
            builder.program_invoke(self._specialist.invoke_node_id)
        builder.await_event(spec.await_node_id, event_ref=spec.event_ref)
        builder.region(self._loop_region_id, "loop")
        builder.return_region(self.RETURN_REGION_ID)
        context_sink = (
            spec.capability_node_id if spec.capability_ref is not None else spec.await_node_id
        )
        self.context_flow(
            ContextEdge(spec.model_node_id, context_sink, self.context_type_ref or "")
        )
        for hook in hooks:
            self.bind_hook(hook)
        builder.model_requirement(spec.model_target_ref)
        if spec.capability_ref is not None:
            builder.capability_requirement(spec.capability_ref)
        source_nodes = [
            (spec.model_node_id, "model.call"),
            *(([(spec.capability_node_id, "capability.invoke")]) if spec.capability_ref is not None else []),
            *(([(self._specialist.new_node_id, "program.new"), (self._specialist.invoke_node_id, "program.invoke")]) if self._specialist is not None else []),
            (spec.await_node_id, "await.event"),
        ]
        for line, (node_id, annotation) in enumerate(source_nodes, start=1):
            builder.node_span(node_id, self._source_file, line, annotation)
        return self.annotate_region(self._loop_region_id, "conversational_loop")
